### ISSUE TITLE
Multi-tenancy (User pool per Tenant) without redundant infrastructure

### ISSUE THREAD

**Issue by taijuten** [supertokens/supertokens-core#283]
## 🚀 Feature
**TL;DR: A single core handling multiple tenants by the way of optional request header**

SuperTokens appears to be a great product, though many SaaS products require a multi-tenancy solution.
One of the most difficult to solve is the "User pool per Tenant" pattern, where users are unique within a tenant. (e.g. bob@domain.com could be registered on Tenant A, Tenant B, or both).

**To an extent, this is currently achievable with ST, with some customisation, by doing the following:**
* Front-end template can be adjusted to submit another form-field for `tenantId` (this could be a hidden form-field for each tenant's front-end)
* Backend `config.js` can be extended to get `appInfo` etc based on the `tenantId` passed in (such as from a separate dynamoDB / data store)
* A core is created per tenant, with the environment variables set to ensure the table name is equal to the `tenantId`
* A shared DB is created (or separate per tenant, if desired)

## Implementation details

In order to reduce redundant infrastructure for this usage pattern, the core should be able to handle requests for a number of tenants, and differentiate between them. An optional request header may cater for this, allowing a tenant to be specified, and switching table appropriately.

**Comment by rishabhpoddar** [comment]
An idea for this (as already discussed on discord): https://discord.com/channels/603466164219281420/603466164219281426/868948314056429598

**Comment by ArpanKIIT2017** [comment]
I can volunteer for this and take part of it, at least backend work I can take up. I myself anyway need it in my new SaaS app.

**Comment by rishabhpoddar** [comment]
WIP: https://github.com/supertokens/supertokens-core/pull/555
